File: core/scala/src/main/scala/org/aiddl/core/scala/function/NamedFunction.scala
package org.aiddl.core.scala.function

import org.aiddl.core.scala.eval.Evaluator
import org.aiddl.core.scala.representation.{Substitution, Sym, Term}

/**
 * Functions created via #def are instantiated using this class.
 * Function arguments are applied via substitution. There are two
 * options:
 *
 * 1. If an argument pattern is provided, it will be matched against function arguments to create a substitution.
 * 2. Otherwise, the term <code>#arg</code> will be replaced by the argument of the function.
 *
 * @param f term representing the function
 * @param eval evaluator used to compute the function
 * @param args argument pattern
 */
class NamedFunction(f: Term, eval: Evaluator, args: Option[Term]) extends Function {
  def apply(x: Term): Term = {
    (args match {
      case Some(pattern) => pattern unify x
      case None => {
        val s = new Substitution
        s.add(Sym("#self"), x)
        s.add(Sym("#arg"), x)
        Some(s)
      }
    }) match {
      case Some(s) => eval(f \ s)
      case None => throw new IllegalArgumentException("Could not match argument " + x + " to " + args)
    }
  }
}

File: core/java/src/main/java/org/aiddl/core/java/function/NamedFunction.java
package org.aiddl.core.java.function;

import org.aiddl.core.java.eval.Evaluator;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.representation.Substitution;
import org.aiddl.core.java.representation.Term;

public class NamedFunction implements Function {
	
	private Term name;	
	private Term f;
	private Term args = null;
	private Evaluator e;
	
	public NamedFunction( Term name, Term f, Evaluator e ) {
		this.name = name;
		this.f = f;
		this.e = e;
	}
	
	public NamedFunction( Term name, Term args, Term f, Evaluator e ) {
		this.name = name;
		this.f = f;
		this.e = e;
		this.args = args;
	}

	@Override
	public Term apply(Term args) {
		Substitution s;
		if ( this.args == null ) {
			s = new Substitution();
			s.add(Term.sym("#self"), args);
			s.add(Term.sym("#arg"), args);
		} else {
			s = this.args.match(args);
		}
		return e.apply(f.substitute(s));
	}
	
	@Override
	public String toString() {
		return f.toString();
	}
}
